[wip] Added bulk add to DC & BO Assets.

xliiv opened the pull request:
Added new field MultilineField which handles seperated values.
Added MultivalueFormMixin which validates many MultilineField in one form.

connected to: #1606

vi4m commented:
screenshotplease.

mkurek commented on src/ralph/admin/fields.py:
```diff
@@ -0,0 +1,109 @@
+                items.append(item.strip())
+        return items
+
+    def clean(self, value):
```
is this method necessary if it only call super?

mkurek commented on src/ralph/admin/fields.py:
```diff
@@ -0,0 +1,109 @@
+    """A form that has several multiline fields that need to have the
+    same number of entries.
+
+    :param multivalue_fields: list of form fields which require the same item
```
describe `one_of_mulitvalue_required` as well

mkurek commented on src/ralph/admin/fields.py:
```diff
@@ -0,0 +1,109 @@
+
+    def __init__(self, allow_duplicates=True, *args, **kwargs):
+        self.allow_duplicates = allow_duplicates
+        super(MultilineField, self).__init__(*args, **kwargs)
```
we should stick to one `super` convention - either `super(MultilineField, self).__init__(*args, **kwargs)` or `super().__init__(*args, **kwargs)`

mkurek commented on src/ralph/admin/fields.py:
```diff
@@ -0,0 +1,109 @@
+    @transaction.atomic
+    def save_model(self, request, obj, form, change):
+        super().save_model(request, obj, form, change)
+        obj = self.model(**form.cleaned_data)
```
why don't you use `obj` passed to this method?

mkurek commented on src/ralph/admin/fields.py:
```diff
@@ -0,0 +1,109 @@
+        obj = self.model(**form.cleaned_data)
+        for idx, form_values in enumerate(self.multivalues_rows()):
+            for value, field_name in zip(form_values, self.multivalue_fields):
+                setattr(obj, field_name, value)
```
shouldn't `super().save_model(request, obj, form, change)` be called there?

mkurek commented on src/ralph/back_office/admin.py:
```diff
@@ -35,6 +48,7 @@ class BackOfficeAssetAdmin(
     list_select_related = ['model', 'user', 'warehouse', 'model__manufacturer']
     raw_id_fields = ['model', 'user', 'owner', 'service_env']
     resource_class = resources.BackOfficeAssetResource
+    form = BackOfficeAssetForm
```
this form should be used only to add asset - not to edit

xliiv replied:
done

mkurek commented on src/ralph/admin/fields.py:
```diff
@@ -0,0 +1,109 @@
+    :param multivalue_fields: list of form fields which require the same item
+    count
+    """
+    multivalue_fields = []
```
wouldn't it be better to define it in form's meta? @ar4s what do you think?

mkurek commented on src/ralph/admin/fields.py:
```diff
@@ -0,0 +1,109 @@
+        self.message_user(request, msg)
+
+    def equal_count_validator(self, cleaned_data):
+        """Adds a validation error if if form's multivalues fields have
```
duplicated `if`

mkurek commented on src/ralph/admin/tests/tests_fields.py:
```diff
@@ -0,0 +1,90 @@
+        self.assertIn('barcode', form.errors)
+
+
+class MultilineFieldTest(SimpleTestCase):
```
could you write additional test when none of required fields is filled (fields have blank values)?

Commit:
Added bulk add to DC & BO Assets.

Added new field MultilineField which handles seperated values.
Added MultivalueFormMixin which validates many MultilineField in one form.

ar4s commented on src/ralph/admin/tests/tests_functional.py:
```diff
@@ -1,4 +1,4 @@
-# -*- coding: utf-8 -*-
+# -*- cmding: utf-8 -*-
```
wut?

ar4s commented on src/ralph/back_office/admin.py:
```diff
@@ -12,6 +14,21 @@
+            self.fields['sn'] = MultilineField('sn')
+            self.fields['barcode'] = MultilineField('barcode')
+            self.multivalue_fields = ['sn', 'barcode']
+            self.one_of_mulitvalue_required = ['sn', 'barcode']
```
`self.Meta.multivalue_fields = ['sn', 'barcode']`
